app/scripts/migrations/127.ts

```diff
@@ -0,0 +1,399 @@
+      if (
+        isObject(networkConfiguration) &&
+        typeof networkConfiguration.chainId === 'string'
+      ) {
+        (acc[networkConfiguration.chainId] ??= []).push(networkConfiguration);
+      }
+      return acc;
+    },
+    {},
+  );
+
+  // Get transaction history in reverse chronological order to help with tie breaks
+  const transactions: RuntimeObject[] =
+    hasProperty(state, 'TransactionController') &&
+    isObject(state.TransactionController) &&
+    Array.isArray(state.TransactionController.transactions)
+      ? state.TransactionController.transactions
+          .filter(
+            (tx) =>
+              isObject(tx) &&
+              typeof tx.time === 'number' &&
+              typeof tx.networkClientId === 'string',
+          )
+          .sort((a, b) => b.time - a.time)
+      : [];
+
+  // For each chain id, merge the array of network configurations
+  const networkConfigurationsByChainId = Object.entries(
+    networkConfigurationArraysByChainId,
+  ).reduce((acc: Record<string, unknown>, [chainId, networks]) => {
+    //
+    // Calculate the tie breaker network, whose values will be preferred
+    let tieBreaker: RuntimeObject | undefined;
+
+    // If one of the networks is the globally selected network, use that
+    tieBreaker = networks.find(
+      (network) => network.id === networkState.selectedNetworkClientId,
+    );
+
+    // Otherwise use the network that was most recently transacted on
+    if (!tieBreaker) {
+      transactions
+        .filter((tx) => tx.chainId === chainId)
+        .some(
+          (tx) =>
+            (tieBreaker = networks.find(
+              (network) => network.id === tx.networkClientId,
+            )),
+        );
+    }
+
+    // If no transactions were found for the chain id,
+    // try to fall back to the built in infura network
+    if (!tieBreaker) {
+      tieBreaker = networks.find((network) => network.type === 'infura');
+    }
+
+    // Calculate the unique set of valid rpc endpoints for this chain id
+    const rpcEndpoints = networks.reduce(
+      (endpoints: RuntimeObject[], network) => {
+        if (network.id && network.rpcUrl) {
+          //
+          // Check if the endpoint is a duplicate, which the network controller doesn't allow
+          const findDuplicate = (endpoint: unknown) =>
+            isObject(endpoint) &&
+            ((typeof endpoint.url === 'string' &&
+              typeof network.rpcUrl === 'string' &&
+              URI.equal(endpoint.url, network.rpcUrl)) ||
+              // This should not be possible, but protect against duplicate network client ids
+              endpoint.networkClientId === network.id);
+
+          // The endpoint must be unique across all chains, not just within each
+          const duplicateWithinChain = endpoints.find(findDuplicate);
+          const duplicateAcrossChains = Object.values(acc)
+            .flatMap((n) => (isObject(n) ? n.rpcEndpoints : []))
+            .find(findDuplicate);
```

Comment:
I suppose we have no way to actually verifying which chainId the rpcEndpoint is actually valid for, but we definitely might be keeping it for the wrong chain here since we just take the first appearance of it as "truth"

Reply:
Not sure there is any better heuristic to tiebreak on, if the same endpoint exists on different chains.  I know the network form will prevent this from happening in the first place, so its an edge case.  But theoretically possible if you had the same URL in different casing, as the controller's validation is stronger now.